crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,262 @@
+        }
+        #[cfg(all(not(test), target_os = "linux"))]
+        if let Some(handle) = &self.monitor_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "windows")]
+        if !self.monitor_handle.0.is_null() {
+            crate::windows::deregister_network_monitor(self.monitor_handle.clone());
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::GetIfAddrs;
+
+    use super::*;
+    use lazy_static::lazy_static;
+    use std::{
+        net::Ipv4Addr,
+        sync::{
+            atomic::{AtomicU8, Ordering},
+            Arc, Mutex as StdMutex,
+        },
+    };
+
+    #[derive(PartialEq, Debug)]
+    enum TestOrder {
+        PathChangeNotification,
+        MultipleMonitorInstances,
+        MonitorDropping,
+        GatherIfError,
+    }
+
+    impl TestOrder {
+        fn next(&mut self) -> Self {
+            println!("Changing order");
+            match self {
+                TestOrder::PathChangeNotification => TestOrder::MultipleMonitorInstances,
+                TestOrder::MultipleMonitorInstances => TestOrder::MonitorDropping,
+                TestOrder::MonitorDropping => TestOrder::GatherIfError,
+                TestOrder::GatherIfError => TestOrder::PathChangeNotification,
+            }
+        }
+    }
+
+    lazy_static! {
+        static ref TEST_MUTEX: Arc<StdMutex<TestOrder>> =
+            Arc::new(StdMutex::new(TestOrder::PathChangeNotification));
+    }
+    static CALL_COUNT: AtomicU8 = AtomicU8::new(0);
+    pub fn gather_local_interfaces(
+        _get_if_addr: GetIfAddrs,
+    ) -> std::io::Result<Vec<if_addrs::Interface>> {
+        let lock = TEST_MUTEX.lock().unwrap();
+        if *lock != TestOrder::GatherIfError {
+            CALL_COUNT.fetch_add(1, Ordering::Relaxed);
+            Ok(vec![if_addrs::Interface {
+                name: "old".to_owned(),
+                addr: if_addrs::IfAddr::V4(if_addrs::Ifv4Addr {
+                    ip: Ipv4Addr::new(10, 0, 0, 1),
+                    netmask: Ipv4Addr::new(255, 255, 255, 0),
+                    broadcast: None,
+                }),
+                index: None,
+                #[cfg(windows)]
+                adapter_name: "{78f73923-a518-4936-ba87-2a30427b1f63}".to_string(),
+            }])
+        } else {
+            Err(std::io::ErrorKind::NotFound.into())
+        }
+    }
+
+    fn test_order_turn(expected_order: TestOrder) {
+        // Acquire the lock and wait for the right turn
+        loop {
+            let lock = TEST_MUTEX.lock().unwrap();
+            if *lock == expected_order {
+                CALL_COUNT.store(0, Ordering::Relaxed);
+                break; // It's this test's turn, proceed
+            }
+            // Release the lock before spinning
+            std::thread::yield_now();
+        }
+    }
+
+    fn increment_order() {
+        let mut lock = TEST_MUTEX.lock().unwrap();
+        *lock = (*lock).next();
+        CALL_COUNT.store(0, Ordering::Relaxed);
+    }
```

Comment:
IMHO this is overengineered..

Locking a single mutex at the beginning of each test is enough:
```
static TESTS_SYNC_MTX: Mutex<()> = Mutex::new(());
    #[tokio::test()]
    async fn test_path_change_notification() {
        let _m = TESTS_SYNC_MTX.lock();
        CALL_COUNT.store(0, Ordering::Relaxed);
       // ...
    }
    #[tokio::test()]
    async fn test_multiple_monitor_instances() {
       let _m = TESTS_SYNC_MTX.lock();
       CALL_COUNT.store(0, Ordering::Relaxed);
       // ...
    }
    ....
```

Reply:
The main problem comes CALL_ONCE on apple monitor, and that BROADCAST_SEND() callback is issued as soon as handler is subscribed which leads to 1 more call than needed. I would like to keep the CALL_ONCE on Apple Monitor. Cancel call requires for me to get the monitor handle, which again complicates things alot because of CALL_ONCE. Looking at it again, it can be simplified, but a simple lock mechanism won't work.